Mathlib/RingTheory/PicardGroup.lean

```diff
@@ -0,0 +1,425 @@
+
+instance : Small.{u} (Skeleton <| ModuleCat.{u} R)ˣ :=
+  let sf := Σ n, Submodule R (Fin n → R)
+  have {M N : sf} : M = N → (_ ⧸ M.2) ≃ₗ[R] _ ⧸ N.2 := by rintro rfl; exact .refl ..
+  let f (M : (Skeleton <| ModuleCat.{u} R)ˣ) : sf :=
+    ⟨_, LinearMap.ker (Finite.exists_fin' R M).choose_spec.choose⟩
+  small_of_injective (f := f) fun M N eq ↦ Units.ext <| Quotient.out_equiv_out.mp
+    ⟨((Finite.reprEquiv R M).symm ≪≫ₗ this eq ≪≫ₗ Finite.reprEquiv R N).toModuleIso⟩
+
+/-- The Picard group of a commutative ring R consists of the invertible R-modules,
+ up to isomorphism. -/
+abbrev CommRing.Pic (R : Type u) [CommRing R] : Type u :=
+  Shrink (Skeleton <| ModuleCat.{u} R)ˣ
+
+open CommRing (Pic)
+
+variable {R} in
+/-- A representative of an element in the Picard group. -/
+abbrev CommRing.Pic.AsModule (M : Pic R) : Type u := ((equivShrink _).symm M).val
+
+noncomputable instance : CoeSort (Pic R) (Type u) := ⟨Pic.AsModule⟩
+
+private noncomputable def equivShrinkLinearEquiv (M : (Skeleton <| ModuleCat.{u} R)ˣ) :
+    equivShrink _ M ≃ₗ[R] M :=
+  have {M N : Skeleton (ModuleCat.{u} R)} : M = N → M ≃ₗ[R] N := by rintro rfl; exact .refl ..
+  this (by simp)
+
+namespace Module.Invertible
+
+variable (M N : Type*) [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+  [Module.Invertible R M] [Module.Invertible R N]
+
+instance : Module.Invertible R (Finite.repr R M) := .congr (Finite.reprEquiv R M).symm
+
+/-- The class of an invertible module in the Picard group. -/
+noncomputable def toPic : Pic R := equivShrink _ <|
+  letI M' := Finite.repr R M
+  .mkOfMulEqOne ⟦.of R M'⟧ ⟦.of R (Dual R M')⟧ <| by
+    rw [← toSkeleton, ← toSkeleton, mul_comm, ← Skeleton.toSkeleton_tensorObj]
+    exact Quotient.sound ⟨(Invertible.linearEquiv R _).toModuleIso⟩
+
+/-- `toPic R M` is indeed the class of `M`. -/
+noncomputable def toPic.linearEquiv : toPic R M ≃ₗ[R] M :=
+  equivShrinkLinearEquiv R _ ≪≫ₗ (Quotient.mk_out (s := isIsomorphicSetoid _)
+    (ModuleCat.of R (Finite.repr R M))).some.toLinearEquiv ≪≫ₗ Finite.reprEquiv R M
+
+variable {R M N}
+
+theorem toPic_eq_iff {N : Pic R} : toPic R M = N ↔ Nonempty (M ≃ₗ[R] N) where
+  mp := (· ▸ ⟨(toPic.linearEquiv R M).symm⟩)
+  mpr := fun ⟨e⟩ ↦ ((equivShrink _).apply_eq_iff_eq_symm_apply).mpr <|
+    Units.ext <| Quotient.mk_eq_iff_out.mpr ⟨(Finite.reprEquiv R M ≪≫ₗ e).toModuleIso⟩
+
+theorem toPic_eq_self {M : Pic R} : toPic R M = M := toPic_eq_iff.mpr ⟨.refl ..⟩
+
+theorem _root_.CommRing.Pic.ext_iff {M N : Pic R} : M = N ↔ Nonempty (M ≃ₗ[R] N) := by
+  rw [← toPic_eq_iff, toPic_eq_self]
+
+theorem toPic_eq_toPic_iff : toPic R M = toPic R N ↔ Nonempty (M ≃ₗ[R] N) :=
+  let eN := toPic.linearEquiv R N
+  toPic_eq_iff.trans ⟨fun ⟨e⟩ ↦ ⟨e ≪≫ₗ eN⟩, fun ⟨e⟩ ↦ ⟨e ≪≫ₗ eN.symm⟩⟩
+
+theorem toPic_self : toPic R R = 1 :=
+  congr_arg (equivShrink _) <| Units.ext <| Quotient.sound ⟨(Finite.reprEquiv R R).toModuleIso⟩
+
+theorem toPic_eq_one_iff : toPic R M = 1 ↔ Nonempty (M ≃ₗ[R] R) := by
```

Comment:
Can we also have a version with `Free` (and then also the backward version with an instance `Free`)? Thanks!

Reply:
I think it's not inconvenient to rewrite by `free_iff_linearEquiv` so I'm not adding another iff statement, but I'll add the backward version with `Free`.